File: src/test_stand.py
"""
test_stand.py: The model for the Test Stand
"""

from labjack import ljm
import threading
from util import *
from typing import List
import time

class TestStand:

    def __init__(self, handle: int) -> None:
        self.sv_states = [False, False, False, False, False]
        self.sv_dio = [1, 0, 2, 3, 4]
        self.sv_freq = 1000
        self.sv_dc = 60
        self.sv_timers = []
        self.handle = handle


        self.mav_states = [False, False]
        self.mav_dio = [5, 6]

        # sensor dict
        self.sensor_dict = self.initialize_sensors()

        # pressure (PSI) - 8x
        self.pt_pressures = []

        # Thermocouplers (Temp F) - 4x
        self.tc_temps = []

        # Flowmeter (Gallons per min) - 1x
        self.flowmeter_gpm = -1

        # Load cell (KLbs) - 2x
        self.load_cell_weights = []

        self.awaited_mode = -1

        self.tanks_purging = False
        self.engine_purging = False
        self.filling = False
        

    def mav_on(self, num: int) -> None:
        if not self.mav_states[num - 1]:
            if num == 1:
                self.mav_states[num - 1] = True
                self._mav_actuate(num, 26.4)
            else:
                return #TODO SCUFFED PWM
                self.mav_states[num - 1] = True
                self._mav_actuate(num, 72.6)

    def mav_off(self, num: int) -> None:
        if self.mav_states[num - 1]:
            if num == 1:
                self.mav_states[num - 1] = False
                self._mav_actuate(num, 72.6)
            else:
                return #TODO SCUFFED PWM
                self.mav_states[num - 1] = False
                self._mav_actuate(num, 26.4)

    def _mav_actuate(self, num: int, dc: float) -> None:
        if not use_labjack:
            return
        
        pwmDIO = self.mav_dio[num - 1]
        roll_value = 242424.24242424243

        config_a = dc * roll_value / 100

        aNames = [
            "DIO_EF_CLOCK0_ROLL_VALUE",
            "DIO_EF_CLOCK0_ENABLE",

            "DIO%i_EF_ENABLE" % pwmDIO,
            "DIO%i_EF_CONFIG_A" % pwmDIO,
            "DIO%i_EF_ENABLE" % pwmDIO,
        ]

        aValues = [
            roll_value,
            1,

            0,
            config_a,
            1
        ]
        numFrames = len(aNames)
        results = ljm.eWriteNames(self.handle, numFrames, aNames, aValues)

    def _sv_pwm(self, num: int) -> None:
        pwmDIO = self.sv_dio[num]

        roll_value = 80_000_000 / self.sv_freq
        config_a = self.sv_dc * roll_value / 100
        
        aNames = [
            "DIO_EF_CLOCK0_ROLL_VALUE",
            "DIO_EF_CLOCK0_ENABLE",

            "DIO%i_EF_ENABLE" % pwmDIO,
            "DIO%i_EF_CONFIG_A" % pwmDIO,
            "DIO%i_EF_ENABLE" % pwmDIO,
        ]

        aValues = [
            roll_value,
            1,

            0,
            config_a,
            1
        ]
        numFrames = len(aNames)
        results = ljm.eWriteNames(self.handle, numFrames, aNames, aValues)

    def sv_on(self, num: int) -> None:
        if not self.sv_states[num - 1]:
            if self.sv_dio[num-1] == 1:
                return #TODO scuffed PWM
            
            self.sv_states[num - 1] = True

            if not use_labjack:
                return
            
            dio = "FIO" + str(self.sv_dio[num - 1])

            ljm.eWriteName(self.handle, dio, 1)
            
            temp_timer = threading.Timer(150 / 1000, self._sv_pwm, [num - 1])
            temp_timer.start()
        
    def sv_off(self, num: int) -> None:
        if self.sv_states[num - 1]:
            self.sv_states[num - 1] = False

            if not use_labjack:
                return
            
            dio = "FIO" + str(self.sv_dio[num - 1])
            aNames = ["DIO%i_EF_ENABLE" % self.sv_dio[num - 1]]
            aValues = [0]
            numFrames = len(aNames)
            results = ljm.eWriteNames(self.handle, numFrames, aNames, aValues)
            ljm.eWriteName(self.handle, dio, 0)

    def set_sv_states(self, states: List[bool]) -> None:
        for i in range(5):
            if states[i]:
                self.sv_on(i+1)
            else:
                self.sv_off(i+1)

    # Modes

    def confirm_mode(self, mode: Mode) -> None:
        if mode == Mode.DEFAULT:
            self.default()
        elif mode == Mode.PREFIRE_PURGE_TANKS:
            self.prefire_purge_tanks()
        elif mode == Mode.PREFIRE_PURGE_ENGINE:
            self.prefire_purge_engine()
        elif mode == Mode.FILL:
            self.fill()
        elif mode == Mode.SUPERCHARGE:
            self.supercharge()
        elif mode == Mode.POSTFIRE_PURGE_ENGINE:
            self.postfire_purge_engine()
        elif mode == Mode.FIRE:
            self.fire()

    def default(self) -> None:
        self.set_sv_states([False] * 5)

        self.mav_off(1)
        self.mav_off(2)
    
    def prefire_purge_tanks(self) -> None:
        if not self.tanks_purging:
            self.set_sv_states([False, True, False, False, True])
        else:
            self.set_sv_states([False] * 5)

    def prefire_purge_engine(self) -> None:
        if not self.engine_purging:
            self.set_sv_states([True, False, False, False, False])
        else:
            self.set_sv_states([False] * 5)

    def fill(self) -> None:
        if not self.filling:
            self.set_sv_states([False, False, True, False, False])
        else:
            self.set_sv_states([False] * 5)

    def supercharge(self) -> None:
        self.set_sv_states([False, False, False, True, False])

    def postfire_purge_engine(self) -> None:
        self.set_sv_states([False, True, False, True, False])
    
    def fire(self) -> None:
        self.set_sv_states([False, False, False, True, False])

        self.mav_on(1)
        self.mav_on(2)
        timer_1 = threading.Timer(150 / 1000, self.mav_off, [1])
        timer_2 = threading.Timer(150 / 1000, self.mav_off, [2])

    # Sensors
    
    def initialize_sensors(self) -> dict:
        """
        Initialize all sensors with their attributes.
        Sensors are initialized as follows:
        s = Sensor(voltage min, voltage max, value 1, value 2)

        Returns a dictionary of initialized sensors with indices as keys.

        TO DO: Add details for reading (positive and negative channels, etc), gain, offset
        """
        pt2000 = PT_FM(0.0, 10.0, 0.0, 2000.0)
        pt3000 = PT_FM(0.0, 10.0, 0.0, 3000.0)
        pt1500 = PT_FM(0.0, 10.0, 0.0, 1500.0)
        tc = TC(0.0, 10.0, 1.0, 200.0, ljm.eReadName(self.handle, "AIN52"))
        fm = PT_FM(1.72, 10.32, 2.5, 29.0)
        lc1000 = LC(0.0, 0.0036, 31.27993035, -0.2654580671) 
        lc2000 = LC(0.0, 0.0036, 60.25906654, -0.02513497142)


        return {0: pt2000, 1: pt2000, 2: pt3000, 3: pt3000, 4: pt1500, 5: pt2000, 6: pt3000, 7: pt2000,
                8: tc, 9: tc, 10: tc,
                11: fm,
                12: lc2000, 13: lc1000}

    

File: src/test_stand.cpp
#include "test_stand.hpp"
#include <chrono>
#include <cstdio>
#include <thread>

TestStand::TestStand(int handle) : handle(handle) {
}

void TestStand::sv_toggle(int num) {
    if (sv_states[num - 1]) {
        sv_off(num);
    } else {
        sv_on(num);
    }
}

void TestStand::sv_on(int num) {
#ifdef USE_LABJACK
    // For SV 4, we do fake PWM
    if (num != 4) {
        char fio_name[5];
        snprintf(fio_name, 5, "FIO%d", sv_dio[num - 1]);

        err = LJM_eWriteName(handle, fio_name, 1);
        ErrorCheckWithAddress(err, error_address, "LJM_eWriteNames");

        std::thread timer_thread([this, num]() {
            this->sv_pwm(num);
        });
        timer_thread.detach();
    } else {
        int queueVals;
        int samplesToWrite = 512;
        
        double* values = new double[samplesToWrite];
        double increment = double(1) / samplesToWrite;
        // Make an arbitrary waveform that increases voltage linearly from 0-2.5V
        for (int i = 0; i < samplesToWrite; i++) {
            double sample = 2.5 * increment * i;
            values[i] = sample;
        }

        err = LJM_WriteAperiodicStreamOut(
            handle,
            0,
            512,
            values,
            &queueVals
        );
    }
#endif
    sv_states[num - 1] = true;
}

void TestStand::sv_off(int num) {
    sv_states[num - 1] = false;

#ifdef USE_LABJACK
    char fio_name[5];
    snprintf(fio_name, 5, "FIO%d", sv_dio[num - 1]);

    char dio_ef_enable[16];
    snprintf(dio_ef_enable, 16, "DIO%d_EF_ENABLE", sv_dio[num - 1]);

    err = LJM_eWriteName(handle, dio_ef_enable, 0);
    ErrorCheckWithAddress(err, error_address, "LJM_eWriteName");

    err = LJM_eWriteName(handle, fio_name, 0);
    ErrorCheckWithAddress(err, error_address, "LJM_eWriteName");
#endif
}

void TestStand::mav_toggle() {
    if (mav_state) {
        mav_off();
    } else {
        mav_on();
    }
}

void TestStand::mav_on() {
    mav_state = true;

#ifdef USE_LABJACK
    mav_pwm(25.0);
#endif
}

void TestStand::mav_off() {
    mav_state = false;

#ifdef USE_LABJACK
    mav_pwm(60.0);
#endif
}

double TestStand::pt_scale(float volt_act, float volt_min, float volt_max, float val_min, float val_max) {
    return ((volt_act - volt_min) * (val_max - val_min)) / (volt_max - volt_min) + val_min;
}

/**
 * Scales thermocouple voltage readings into a farenheit value.
 */
double TestStand::tc_scale(float volt_act) {
    double r_ref = 15000;
    double alpha = 0.00392;
    double r_nom = 100;
    double value;
    double dac_ref = LJM_eReadName(handle, "AIN52", &value);
    double r_rtd = (r_ref * (dac_ref - volt_act)) / (volt_act);
    double delta_tempC = (1 / alpha) * ((r_rtd / r_nom) - 1);
    return delta_tempC * (9 / 5) + 32;
}

/**
 * Scales load cell voltage readings into a pounds value.
 */
double TestStand::lc_scale(float volt_act, float m, float b) {
    return 1000 * ((m * volt_act) + b);
}

void TestStand::mav_pwm(float dc) {
    float config_a = dc * MAV_ROLL_VALUE / 100;

    char dio_ef_enable[15];
    snprintf(dio_ef_enable, 15, "DIO%d_EF_ENABLE", MAV_DIO);

    char dio_ef_config_a[17];
    snprintf(dio_ef_config_a, 17, "DIO%d_EF_CONFIG_A", MAV_DIO);

    const int NUM_FRAMES = 5;
    const char *names[NUM_FRAMES] = {"DIO_EF_CLOCK0_ROLL_VALUE", "DIO_EF_CLOCK0_ENABLE", dio_ef_enable, dio_ef_config_a, dio_ef_enable};
    double values[NUM_FRAMES] = {MAV_ROLL_VALUE, 1, 0, config_a, 1};

    err = WriteNames(handle, NUM_FRAMES, names, values, &error_address);
    ErrorCheckWithAddress(err, error_address, "LJM_eWriteNames");
}

void TestStand::sv_pwm(int num) {
    std::this_thread::sleep_for(std::chrono::milliseconds(150));

    int pwm_dio = sv_dio[num - 1];
    float config_a = SV_DC * SV_ROLL_VALUE / 100;

    char dio_ef_enable[15];
    snprintf(dio_ef_enable, 15, "DIO%d_EF_ENABLE", pwm_dio);

    char dio_ef_config_a[17];
    snprintf(dio_ef_config_a, 17, "DIO%d_EF_CONFIG_A", pwm_dio);

    const int NUM_FRAMES = 5;
    const char *names[NUM_FRAMES] = {"DIO_EF_CLOCK0_ROLL_VALUE", "DIO_EF_CLOCK0_ENABLE", dio_ef_enable, dio_ef_config_a, dio_ef_enable};
    double values[NUM_FRAMES] = {SV_ROLL_VALUE, 1, 0, config_a, 1};

    err = WriteNames(handle, NUM_FRAMES, names, values, &error_address);
    ErrorCheckWithAddress(err, error_address, "LJM_eWriteNames");
}

void TestStand::set_sv_states(std::string mask) {
    for (int i = 0; i < mask.size(); i++) {
        if (mask[i] - '0') {
            sv_on(i + 1);
        } else {
            sv_off(i + 1);
        }
    }
}

void TestStand::to_mode(Mode mode) {
    switch (mode) {
    case Mode::prefire_purge_tanks:
        if (!tanks_purging) {
            set_sv_states("01001");
        } else {
            set_sv_states("00000");
        }
        break;
    case Mode::prefire_purge_engine:
        if (!engine_purging) {
            set_sv_states("10000");
        } else {
            set_sv_states("00000");
        }
        break;
    case Mode::fill:
        if (!filling) {
            set_sv_states("00100");
        } else {
            set_sv_states("00000");
        }
        break;
    case Mode::close_vent:
        set_sv_states("00110");
        break;
    case Mode::fire:
        set_sv_states("00000");
        mav_on();
        break;
    case Mode::postfire_purge_engine:
        set_sv_states("10000");
        break;
    default:
        set_sv_states("00000");
        mav_off();
    }

    this->mode = mode;
}